networking/src/Node/Internal.hs

```diff
@@ -665,21 +675,34 @@ startNode logTrace packing peerData mkNodeEndPoint mkReceiveDelay mkConnectDelay
 
 -- | Stop a 'Node', closing its network transport and end point.
 stopNode :: Node packingType peerData -> IO ()
-stopNode Node {..} = do
-    modifyMVar nodeState $ \nodeState ->
+stopNode node = do
+    modifyMVar (nodeState node) $ \nodeState ->
         if _nodeStateClosed nodeState
         then throwIO $ userError "stopNode : already stopped"
         else pure (nodeState { _nodeStateClosed = True }, ())
     -- This eventually will shut down the dispatcher thread, which in turn
     -- ought to stop the connection handling threads.
     -- It'll also close all TCP connections.
-    nodeCloseEndPoint
+    nodeCloseEndPoint node
     -- Must wait on any handler threads. The dispatcher thread will eventually
     -- see an event indicating that the end point has closed, after which it
     -- will wait on all running handlers. Since the end point has been closed,
     -- no new handler threads will be created, so this will block indefinitely
     -- only if some handler is blocked indefinitely or looping.
-    wait nodeDispatcherThread
+    wait (nodeDispatcherThread node)
+    waitForRunningHandlers node
+
+-- | Kill a 'Node', terminating its dispatcher thread, closing its endpoint,
+-- and killing all of its handlers.
+killNode :: Node packingType peerData -> IO ()
+killNode node = do
+    modifyMVar (nodeState node) $ \nodeState ->
+        if _nodeStateClosed nodeState
+        then throwIO $ userError "killNode : already killed"
+        else pure (nodeState { _nodeStateClosed = True }, ())
+    nodeCloseEndPoint node
```

Comment:
Maybe it's nicer to first kill the dispatcher thread and running handlers and close the endpoint afterwards?

Reply:
It's not clear whether it makes any difference. In practice, with the TCP transport, closing the endpoint means closing the socket for each connected peer and then updating some state. This could take a bit of time and I suppose that we shouldn't have the dispatcher thread or any handlers running while it's done. I'll make the change, should be fine.